sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,437 @@
+   */
+  public static boolean fitsOptimizedConversion(
+      CudfUnsafeRow cudfUnsafeRow) {
+    return !cudfUnsafeRow.isVariableWidthSchema() &&
+        cudfUnsafeRow.getVariableSizeDataOffset() < JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED;
+  }
+
+  /**
+   * @see #fitsOptimizedConversion(CudfUnsafeRow)
+   * In some situations, we can calculate an estimate size for a schema withoput constructing
+   * the object yet. In that case, we can use {@link JCudfRowVisitor} which already processed the
+   * schema.
+   */
+  public static boolean fitsOptimizedConversion(
+      JCudfRowOffsetsEstimator cudfRowOffsetMock) {
+    return !cudfRowOffsetMock.hasVarSizeData() &&
+        cudfRowOffsetMock.getCurrentByteOffset() < JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED;
+  }
+
+  /**
+   * A helper class to calculate the columns, validity, and variable width offsets.
+   * This helper is used to get an estimate size for the row including variable-sized rows.
+   * The implementation does not make any assumptions whether the schema is packed or not.
+   * Unlike {@link JCudfRowVisitor}, this implementation does not cache any calculations.
+   */
+  public static class JCudfRowOffsetsEstimator {
+    Attribute[] attributes;
+    int byteCursor = 0;
+    int varSizeColIndex;
+    int validityBytesOffset = 0;
+
+    private JCudfRowOffsetsEstimator(Attribute[] attrArr) {
+      this.attributes = attrArr;
+      resetMeta();
+    }
+
+    private int calcColOffset(int ind) {
+      int res;
+      Attribute attr = attributes[ind];
+      DataType dataType = attr.dataType();
+      DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+      if (isFixedLength(dataType)) {
+        int length = rapidsType.getSizeInBytes();
+        res = alignOffset(byteCursor, length);
+        byteCursor = res + length;
+        return res;
+      }
+      if (!hasVarSizeData()) {
+        // set the index to the first variable size column.
+        varSizeColIndex = ind;
+      }
+      res = alignOffset(byteCursor, JCUDF_VAR_LENGTH_FIELD_ALIGNMENT);
+      byteCursor = res + JCUDF_VAR_LENGTH_FIELD_SIZE;
+      return res;
+    }
+
+    /**
+     * Sets the column offsets
+     * @param offsets array containing the offset for all columns
+     * @return the current byte offset of the row.
+     */
+    public int setColumnOffsets(int[] offsets) {
+      resetMeta();
+      for (int i = 0; i < attributes.length; i++) {
+        offsets[i] = calcColOffset(i);
+      }
+      return byteCursor;
+    }
+
+    private void resetMeta() {
+      varSizeColIndex = attributes.length;
+      byteCursor = 0;
+    }
+
+    private void setValidityBits(int fieldsCount){
+      validityBytesOffset = byteCursor;
+      byteCursor += calculateBitSetWidthInBytes(fieldsCount);
+    }
+
+    public int getEstimateSize() {
+      resetMeta();
+      for (int i = 0; i < attributes.length; i++) {
+        calcColOffset(i);
+      }
```

Comment:
This seems odd to throw away the offsets you're calculating from the return value, but now I see the goal is the side-effect in the function of setting `varSizeColIndex`. I wouldn't have thought much of this had this called `setCOlumnOffsets` and I would have missed the side-effect entirely. I wonder if something should change around here to better name, indicate the side-effect, or not require offset calculation at all if it has been done before.

Reply:
done! changed the name to `advanceColCursorAndGet`